### ISSUE TITLE
vmware_guest: "shutdownguest" does not guarantee state

### ISSUE THREAD

**Issue by karniemi** [ansible/ansible#28498]
##### ISSUE TYPE

 - Bug Report

##### COMPONENT NAME

vmware_guest

##### ANSIBLE VERSION

```
ansible 2.3.1.0-3.el7
```

##### OS / ENVIRONMENT

RHEL7.3

##### SUMMARY

vmware_guest: "shutdownguest" does not guarantee state. We had two tasks in  a playbook:
1) vmware_guest: (state=shutdownguest)
2) vmware_guest_snapshot: (state=absent)

task 2 failed sometimes for "The operation is not allowed in the current state". 

Analysis:
-----------
It would seem that vmware_guest: (state=shutdownguest) does not guarantee that the VM is actually in shutdown state before letting the playbook continue. This causes next tasks in the playbook to fail sometimes.

We tried to work around this bug by waiting VM shutdown. Steps:
1) vmware_guest: (state=shutdownguest)
2) vsphere_guest(vmware_guest_facts=yes), until hw_power_status == "POWERED OFF"
3) vmware_guest_snapshot: (state=absent)

...and we got rid of failures. ( task 2, cloned from [here](https://groups.google.com/forum/#!topic/Ansible-project/vjIl3_L9GW0) )

##### STEPS TO REPRODUCE

See: Summary

```yaml

```

##### EXPECTED RESULTS

vmware_guest: "shutdownguest" returns only after VM is actually shutdown, or fails for error if VM does not shutdown.
##### ACTUAL RESULTS

vmware_guest: "shutdownguest" returns when the shutdown operation is still being executed.

```

```

**Comment by ansibot** [comment]
cc @akasurde @dagwieers @dav1x @jctanner @nerzhul
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by dav1x** [comment]
@karniemi I recreated this as well. This is a design question more than a bug I think. Perhaps a wait_for_shutdown param like wait_for_ip? I think that all of the power states are async in nature to accommodate for speed. 

See: 

https://github.com/vmware/pyvmomi/blob/master/docs/vim/VirtualMachine.rst

Specifically:

ShutdownGuest():
Issues a command to the guest operating system asking it to perform a clean shutdown of all services. Returns immediately and does not wait for the guest operating system to complete the operation.

**Comment by dagwieers** [comment]
We can actually implement it the way we like if needed.

The modules now can fail if the VM is in a state where it cannot perform a request, and we should probably have a retry-mechanism, or a way to make an operation blocking.

**Comment by karniemi** [comment]
@dav1x I have pretty much taken ansible in use due the promise of describing desired system stage using idempotent modules. For any other module by so far, they either have set the desired state before letting the playbooks continue, or they fail in error.  This provides a consistent base for the next actions in the playbook - and I thought that was a consistent design decision for ansible and all the ansible modules. For async purposes, ansible already provides a common concept for all modules: [playbooks_async](http://docs.ansible.com/ansible/latest/playbooks_async.html).

Of course, this module could play around the concept of "state" by defining a "transition state" like "guest_shutdown_triggered"...but I really hope the modules would not start going that way -because it would harm developing robust ansible playbooks.

**Comment by ansibot** [comment]
cc @pdellaert
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by Akasurde** [comment]
@karniemi Would you like to take a look into #31669 as this solves your purpose?

**Comment by Akasurde** [comment]
resolved_by_pr #31669

**Comment by Borrelworst** [comment]
Hi,

I am still experiencing this issue when using the 'state: shutdownguest'.  
In my case I want to shutdown the guest to update the hardware, which fails with the error "The operation is not allowed in the current state."

Ansible Version: 2.7.0

Guest:
OS: RHEL 7.5
VMWareTools: open-vm-tools-10.1.10-3
ESXi Version: 6.5

**Comment by karniemi** [comment]
...it seems that this was implemented as an opt-in bugfix. You need to set state_change_timeout parameter for the module -otherwise the fix won't work. You might want to raise a new issue that the module should wait by default.

**Comment by Borrelworst** [comment]
Thanks for the answer! I did not get that this was the setting you needed to use out of this thread or the PR.
